File: guests/python/grid.py
"""The grid conformance scene, Python port. See
guests/rust/grid.rs and tools/scenes/grid.steps."""

import sys

import kaya

app = kaya.App()

with app.window(title="grid"):
    with kaya.column():
        with kaya.grid(2):
            kaya.label(text="Name:")  # label#0
            kaya.label(text="Ada Lovelace")  # label#1
            kaya.label(text="Role:")  # label#2
            kaya.label(text="Engine programmer")  # label#3
        with kaya.row(grow=1.0):
            kaya.button("left")  # button#0
            kaya.spacer()
            kaya.button("right")  # button#1

sys.exit(app.run())

File: guests/haskell/grid.hs
-- The grid conformance scene, Haskell port. See
-- guests/rust/grid.rs and tools/scenes/grid.steps.

import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  _ <- buildTx app $ do
    windowTitle "grid"
    root <-
      column
        []
        [ gridOf 2
            [ labelText "Name:", -- label#0
              labelText "Ada Lovelace", -- label#1
              labelText "Role:", -- label#2
              labelText "Engine programmer" -- label#3
            ],
          row
            [Grow 1]
            [ buttonOn "left" (return ()), -- button#0
              spacer,
              buttonOn "right" (return ()) -- button#1
            ]
        ]
    mount root
    return ()
  return ()
